File: core/ftr/certificate.py
"""L6 — the statutory output.

Section 63 of the Bharatiya Sakshya Adhiniyam 2023 replaced section 65B of the
Indian Evidence Act with effect from 1 July 2024. Its certificate comes in two
parts — one completed by the person in charge of the device, one by an expert —
and it must state **the hash value of the record and the algorithm used**.

The app computed exactly that at the moment of capture. So Part A can be
pre-populated the instant a record is sealed, which is what makes this layer
nearly free once L4 and L5 exist, and is the part of the system nobody else
builds.

Three rules this module will not bend:

1. **The app never signs for anyone.** Every field a human must attest is emitted
   empty and marked. An auto-filled signature line is a forgery mechanism.
2. **Nothing is asserted that the record does not carry.** If a field has no
   value, it is emitted as absent rather than guessed.
3. **Unverified statutory labels produce a DRAFT.** The field labels in
   ``data/bsa63_schedule.json`` are paraphrases until somebody transcribes the
   Schedule from the bare Act. Until then every certificate says on its face that
   it must not be filed.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .record import SealedRecord

__all__ = ["Schedule", "Certificate", "build_certificate", "SCHEDULE_PATH"]

SCHEDULE_PATH = Path(__file__).parent / "data" / "bsa63_schedule.json"

DRAFT_BANNER = [
    "*** DRAFT — NOT FOR FILING ***",
    "",
    "The statutory field labels below are UNVERIFIED PARAPHRASES. They have not",
    "been transcribed from the bare Act and are very likely wrong in wording, and",
    "possibly in numbering and part assignment.",
    "",
    "Before any certificate is filed, transcribe the Schedule to section 63 from",
    "the Act itself into ftr/data/bsa63_schedule.json and set verified = true.",
    "The computed values below (hash, algorithm, device, period) are correct and",
    "come from the signed record; only the LABELS are provisional.",
]


@dataclass(frozen=True)
class Schedule:
    """The certificate's field structure, loaded from data."""

    raw: dict[str, Any]

    @classmethod
    def load(cls, path: Path = SCHEDULE_PATH) -> "Schedule":
        return cls(json.loads(path.read_text()))

    @property
    def verified(self) -> bool:
        return bool(self.raw.get("verified"))

    @property
    def statute(self) -> str:
        return self.raw.get("statute", "unknown statute")

    def part(self, name: str) -> dict[str, Any]:
        return self.raw[name]

    def fields(self, part: str) -> list[dict[str, str]]:
        return self.part(part)["fields"]


@dataclass(frozen=True)
class Certificate:
    """A populated certificate. Never final, never signed, by construction."""

    schedule: Schedule
    values: dict[str, str]              # machine-filled values, by field key
    blanks: list[str]                   # field keys a human must complete
    record_digest: str
    is_draft: bool

    @property
    def status(self) -> str:
        return "DRAFT — NOT FOR FILING" if self.is_draft else "READY FOR SIGNATURE"

    def text(self, width: int = 78) -> str:
        out: list[str] = []
        if self.is_draft:
            out += DRAFT_BANNER + ["", "=" * width, ""]

        out.append(f"CERTIFICATE UNDER {self.schedule.statute.upper()}")
        out.append("=" * width)
        out.append("")

        for part in ("part_a", "part_b"):
            meta = self.schedule.part(part)
            title = part.replace("_", " ").upper()
            out.append(f"{title} — {meta['signatory_role']}")
            out.append("-" * width)
            for f in self.schedule.fields(part):
                key, label = f["key"], f["label"]
                if key in self.values:
                    value = self.values[key]
                    out.append(f"  {label}:")
                    for line in _wrap(value, width - 6):
                        out.append(f"      {line}")
                else:
                    marker = "[ to be completed by hand ]"
                    out.append(f"  {label}:")
                    out.append(f"      {marker}")
            out.append("")

        out.append("-" * width)
        out.append(f"Status: {self.status}")
        out.append(f"Fields awaiting a human signatory: {len(self.blanks)}")
        out.append("")
        out.append("This certificate accompanies a PRESUMPTIVE field test. A presumptive")
        out.append("test is a screening indication only. It does not identify a substance")
        out.append("and does not replace laboratory analysis.")
        return "\n".join(out)


def _wrap(s: str, w: int) -> list[str]:
    words, line, out = s.split(), "", []
    for word in words:
        if line and len(line) + 1 + len(word) > w:
            out.append(line)
            line = word
        else:
            line = f"{line} {word}".strip()
    if line:
        out.append(line)
    return out or [""]


def build_certificate(rec: SealedRecord, schedule: Schedule | None = None) -> Certificate:
    """Populate Part A from a sealed record. Part B is left entirely to the expert."""
    schedule = schedule or Schedule.load()
    body = rec.body

    device = body.get("device", {})
    operator = body.get("operator", {})
    captured = body.get("captured_at", {})
    att = rec.attestation or {}

    device_bits = [
        f"Android device, verified boot {device.get('verified_boot_state', 'UNKNOWN')}",
        f"bootloader {device.get('bootloader_state', 'UNKNOWN')}",
        f"OS patch level {device.get('os_patch_level', 'unknown')}",
        f"signing key security level {att.get('security_level', 'UNKNOWN')}",
    ]

    values: dict[str, str] = {
        "device_particulars": "; ".join(device_bits),
        "record_produced": f"Field Test Record {body.get('record_uuid', 'unknown')}, "
                           f"sequence {body.get('sequence', '?')} on this device's ledger",
        "hash_value": rec.digest.hex(),
        "hash_algorithm": "SHA-256 over the canonical CBOR encoding of the record",
    }

    if operator.get("id"):
        values["device_operator"] = str(operator["id"])
    if captured.get("device_clock"):
        # One record, one moment. Claiming a "period of regular use" the record does
        # not evidence would be asserting something the device never observed.
        values["period_of_use"] = (
            f"This record was created at {captured['device_clock']} (device clock). "
            f"The device clock is not independently corroborated; the record's position "
            f"in the append-only ledger bounds when it was created."
        )

    blanks = [f["key"] for part in ("part_a", "part_b")
              for f in schedule.fields(part) if f["key"] not in values]

    return Certificate(
        schedule=schedule,
        values=values,
        blanks=blanks,
        record_digest=rec.digest.hex(),
        is_draft=not schedule.verified,
    )

File: app/lib/src/certificate.dart
/// L6 on the device — the statutory certificate and the CCTNS-2.0 envelope.
///
/// A Dart counterpart to `core/ftr/certificate.py` and `esakshya.py`, so an
/// officer can produce the certificate at the point of seizure rather than
/// waiting for someone to run a Python CLI over the record later.
///
/// The three rules from the Python are kept exactly, because they are the whole
/// reason this layer is defensible:
///
///  1. **The app never signs for anyone.** Every field a human must attest is
///     emitted blank and marked. The Schedule's declarations begin "I do hereby
///     solemnly affirm" — precisely the things no program may assert.
///  2. **Nothing is asserted that the record does not carry.** A field the record
///     cannot supply is named as missing rather than left silently blank.
///  3. **An unverified Schedule can only produce a DRAFT.**
library;

import 'dart:convert';

import 'package:flutter/services.dart' show rootBundle;
import 'package:ftr_verify/ftr_verify.dart' as ftr;

class Certificate {
  final Map<String, dynamic> schedule;
  final Map<String, String> values;
  final List<String> blanks;
  final List<String> missingFromRecord;
  final bool isDraft;
  final String digestHex;

  const Certificate({
    required this.schedule,
    required this.values,
    required this.blanks,
    required this.missingFromRecord,
    required this.isDraft,
    required this.digestHex,
  });

  String get status => isDraft ? 'DRAFT — NOT FOR FILING' : 'READY FOR SIGNATURE';

  String get verificationLevel =>
      schedule['verification_level'] as String? ?? 'paraphrase';

  List<Map<String, dynamic>> get parts =>
      (schedule['parts'] as List).cast<Map<String, dynamic>>();
}

class ScheduleLoader {
  static Map<String, dynamic>? _cached;

  /// The Schedule, bundled as an asset from `core/ftr/data/`. One source of
  /// truth stays in Python; `test/schedule_sync_test.dart` fails if it drifts.
  static Future<Map<String, dynamic>> load() async {
    if (_cached != null) return _cached!;
    final raw = await rootBundle.loadString('assets/bsa63_schedule.json');
    return _cached = jsonDecode(raw) as Map<String, dynamic>;
  }
}

Certificate buildCertificate(
    ftr.SealedRecord rec, Map<String, dynamic> schedule) {
  final body = rec.body;
  final device = (body['device'] as Map?) ?? const {};
  final att = rec.attestation;

  final values = <String, String>{
    // The Schedule's own tick-list. A phone is "Mobile".
    'device_type': 'Mobile',
    // Both parts require the hash and the algorithm, and SHA256 is one of the
    // algorithms the Schedule names on its face.
    'hash_value': ftr.hex(rec.digest),
    'hash_value__algorithm': 'SHA256',
  };

  void put(String key, Object? v) {
    if (v != null && '$v'.isNotEmpty) values[key] = '$v';
  }

  put('make_model', device['make_model']);
  put('serial_number', device['serial_number']);
  put('device_identifier', device['device_identifier']);

  values['other_device_information'] = [
    'Field Test Record ${body['record_uuid'] ?? 'unknown'}',
    'sequence ${body['sequence'] ?? '?'} on this device\'s append-only ledger',
    'signing key security level ${att['security_level'] ?? 'UNKNOWN'}',
    'attestation chain ${(att['cert_chain'] as List?)?.length ?? 0} certificate(s)',
    'measured by ${body['pipeline'] ?? 'unknown'} pipeline',
  ].join('; ');

  // Statutory fields the record is expected to supply but cannot. Named rather
  // than left blank, so the gap is not discovered by a magistrate.
  const expected = {
    'make_model': 'Make & Model of the device',
    'serial_number': 'Serial Number of the device',
    'device_identifier': 'IMEI/UIN/UID/MAC/Cloud ID',
  };
  final missing = <String>[];
  expected.forEach((k, label) {
    if (!values.containsKey(k)) missing.add(label);
  });

  final blanks = <String>[];
  for (final part in (schedule['parts'] as List).cast<Map<String, dynamic>>()) {
    for (final item in (part['items'] as List).cast<Map<String, dynamic>>()) {
      final key = item['key'] as String;
      if (!values.containsKey(key)) blanks.add(key);
    }
  }

  return Certificate(
    schedule: schedule,
    values: values,
    blanks: blanks,
    missingFromRecord: missing,
    isDraft: schedule['verified'] != true,
    digestHex: ftr.hex(rec.digest),
  );
}

/// The CCTNS-2.0 handoff envelope.
///
/// Deliberately not a parallel evidence store: CCTNS-2.0 stays the system of
/// record. Field names are PROVISIONAL — no published ingest specification has
/// been found — and the envelope says so about itself.
Map<String, Object?> buildEnvelope(ftr.SealedRecord rec, {String? certificateStatus}) {
  final body = rec.body;
  final ndps = (body['ndps'] as Map?) ?? const {};
  final cls = (body['classification'] as Map?) ?? const {};
  final col = (body['colorimetry'] as Map?) ?? const {};
  final live = (body['liveness'] as Map?) ?? const {};
  final loc = (body['location_bundle'] as Map?) ?? const {};

  return {
    'envelope_version': 1,
    'interface_status': 'PROVISIONAL — field names are not taken from a published '
        'eSakshya/CCTNS-2.0 ingest specification.',
    'produced_by': 'SIH26231 field companion',
    'system_of_record': 'CCTNS-2.0',
    'routing': {
      'fir_reference': ndps['fir_reference'],
      'seizure_memo_ref': ndps['seizure_memo_ref'],
      'sample_ids': ndps['sample_ids'] ?? const [],
    },
    'record': {
      'record_uuid': body['record_uuid'],
      'sequence': body['sequence'],
      'hash_value': ftr.hex(rec.digest),
      'hash_algorithm': 'SHA-256',
      'signature_algorithm': 'ECDSA-P256',
      'key_security_level': rec.attestation['security_level'],
      'pipeline': body['pipeline'],
    },
    'result': {
      'presumptive': true,
      'confirmatory': false,
      'label': cls['label'],
      'prediction_set': cls['prediction_set'] ?? const [],
      'alpha_x1000': cls['alpha_x1000'],
      'gate_passed': col['gate_passed'],
      'refusals': col['refusals'] ?? const [],
      'liveness_checked': live['checked'],
      'liveness_live': live['live'],
    },
    // Carried forward deliberately: an envelope that dropped the anomalies would
    // hand the receiving system a cleaner story than the record tells.
    'location': {
      'channels_agreeing': loc['corroboration_channels_agreeing'],
      'channels_total': loc['corroboration_channels_total'],
      'anomalies': loc['spoof_indicators'] ?? const [],
    },
    if (certificateStatus != null)
      'certificate': {
        'statute': 'Bharatiya Sakshya Adhiniyam 2023, section 63',
        'status': certificateStatus,
      },
    'caveat': 'Presumptive field test. A screening indication only: it does not '
        'identify a substance, does not establish quantity, and does not replace '
        'laboratory analysis under NDPS procedure.',
  };
}

/// Render the certificate as plain text, for the handoff bundle.
///
/// Plain text on purpose: the certificate is filed on paper and signed by hand,
/// and a format nobody needs a viewer for is one nobody can claim was altered
/// in rendering. Blanks are printed as blanks — the whole point of this layer is
/// that the app never signs for a human — and anything the record could not
/// supply is named rather than silently omitted.
String renderCertificate(Certificate c) {
  final b = StringBuffer()
    ..writeln('CERTIFICATE UNDER SECTION 63, BHARATIYA SAKSHYA ADHINIYAM 2023')
    ..writeln('=' * 62)
    ..writeln()
    ..writeln('STATUS: ${c.status}')
    ..writeln('Schedule text verification level: ${c.verificationLevel}');
  if (c.isDraft) {
    b
      ..writeln()
      ..writeln('This is a DRAFT. The Schedule text bundled with this app has not')
      ..writeln('been checked against the official eGazette publication, so nothing')
      ..writeln('here may be filed until it has been.');
  }
  if (c.missingFromRecord.isNotEmpty) {
    b
      ..writeln()
      ..writeln('NOT SUPPLIED BY THE RECORD — must be completed by hand:');
    for (final m in c.missingFromRecord) {
      b.writeln('  - $m');
    }
  }

  for (final part in c.parts) {
    b
      ..writeln()
      ..writeln('-' * 62)
      ..writeln('${part['title']}')
      ..writeln('-' * 62);
    final preamble = part['preamble'] as String?;
    if (preamble != null && preamble.isNotEmpty) {
      b
        ..writeln()
        ..writeln(preamble);
    }
    b.writeln();
    for (final item in (part['items'] as List).cast<Map<String, dynamic>>()) {
      final key = item['key'] as String;
      final label = item['label'] as String? ?? key;
      final value = c.values[key];
      b.writeln('  $label:'.padRight(46) +
          (value ?? '________________________  [to be completed]'));
    }
  }

  b
    ..writeln()
    ..writeln('-' * 62)
    ..writeln('Signature: ______________________   Date: ______________')
    ..writeln()
    ..writeln('Record digest (SHA256): ${c.digestHex}')
    ..writeln()
    ..writeln('A presumptive field test is a screening indication. It does not')
    ..writeln('identify a substance and does not replace laboratory analysis.');
  return b.toString();
}
